### PR TITLE
feat(rpc): introduce getblocktemplate rpc call with stub fields

### LINKED ISSUE #5451: Introduce `getblocktemplate` RPC method
Tasks:

- [ ] Add a `getblocktemplate` RPC method behind the `getblocktemplate-rpcs` rust feature.
- [ ] The method should have no arguments.
- [ ] Create the response structure `GetBlockTemplate` with fields: https://github.com/ZcashFoundation/zebra/issues/5235#issuecomment-1285174731
- [ ] Always return an empty `GetBlockTemplate`
- [ ] Test the rpc method is callable and returns a `GetBlockTemplate` type.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

This is pretty much what i have in mind for https://github.com/ZcashFoundation/zebra/issues/5451

It introduces the rpc call and all the fields using mostly strings as placeholders. All data will be filled in the other tickets, this is kind of the skeleton.

This is a draft, feel free to suggest changes. 

## Solution

Introduce `getblocktemplate` rpc methjod, return empty `GetblockTemplate` struct on call.

### Reviewer Checklist

  - [x] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [x] Are the PR labels correct?
  - [x] Does the code do what the ticket and PR says?
  - [x] How do you know it works? Does it have tests?

## Follow Up Work

- https://github.com/ZcashFoundation/zebra/issues/5452
- https://github.com/ZcashFoundation/zebra/issues/5455
- https://github.com/ZcashFoundation/zebra/issues/5453
- https://github.com/ZcashFoundation/zebra/issues/5454

### CODE REVIEW

File: zebra-rpc/src/methods/get_block_template.rs [hunk 1]
```diff
@@ -139,8 +150,106 @@ where
...
+                finalsaplingroothash: empty_string.clone(),
+                defaultroots: DefaultRoots {
+                    merkleroot: empty_string.clone(),
+                    chainhistoryroot: empty_string.clone(),
+                    authdataroot: empty_string.clone(),
+                    blockcommitmentshash: empty_string.clone(),
+                },
+                transactions: vec![],
+                coinbasetxn: Coinbase {},
+                longpollid: empty_string.clone(),
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template.rs]: Zebra doesn't support the "long polling" feature, so we shouldn't include the `longpollid` field.
(It's an optional `getblocktemplate` feature.)

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good, but I think `longpollid` is out of scope for now.

File: zebra-rpc/src/methods/get_block_template.rs (hunk 1 above)
Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template.rs]: Here's the spec:
https://en.bitcoin.it/wiki/BIP_0022#Optional:_Long_Polling

File: zebra-rpc/src/methods/get_block_template.rs [hunk 2]
```diff
@@ -139,8 +150,106 @@ where
...
+    /// Add documentation.
+    pub mutable: Vec<String>,
+    /// Add documentation.
+    pub noncerange: String,
+    /// Add documentation.
+    pub sigoplimit: u32,
+    /// Add documentation.
+    pub sizelimit: u32,
+    /// Add documentation.
+    pub curtime: u32,
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template.rs]: Optional: For some other RPCs, we've used the Rust standard where words are separated by underscores, then renamed the RPC when serializing.

Did you want to do that here?

https://github.com/ZcashFoundation/zebra/blob/a2a75059d39dcd9aaabb9609a84e0dc59b966535/zebra-rpc/src/methods.rs#L1009

File: zebra-rpc/src/methods/get_block_template.rs [hunk 3]
```diff
@@ -139,8 +150,106 @@ where
...
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct DefaultRoots {
+    /// Add documentation.
+    pub merkleroot: String,
+    /// Add documentation.
+    pub chainhistoryroot: String,
+    /// Add documentation.
+    pub authdataroot: String,
+    /// Add documentation.
+    pub blockcommitmentshash: String,
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template.rs]: Optional: same here, do we want to use underscores and renames?

**Review by teor2345 — COMMENTED** [review]
I wonder if we want to use underscores to match the rest of the RPC implementations.

File: zebra-rpc/src/methods/get_block_template.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template.rs]: That is one less to worry about, thanks! I removed it at https://github.com/ZcashFoundation/zebra/pull/5462/commits/5756a77f89d8597ce146da071878d978c4706711

File: zebra-rpc/src/methods/get_block_template.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template.rs]: Yea, we definitely want to do this. https://github.com/ZcashFoundation/zebra/pull/5462/commits/5756a77f89d8597ce146da071878d978c4706711
Thanks.

File: zebra-rpc/src/methods/get_block_template.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template.rs]: https://github.com/ZcashFoundation/zebra/commit/5756a77f89d8597ce146da071878d978c4706711

File: zebra-rpc/src/methods/get_block_template.rs [hunk 4]
```diff
@@ -139,8 +150,118 @@ where
...
+                bits: empty_string,
+                height: 0,
+            })
+        }
+        .boxed()
+    }
+}
+/// Documentation to be added after we document all the individual fields.
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct GetBlockTemplate {
```
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template.rs]: I think it would be nice to separate the types for each method into their own files.

Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template.rs]: I think this is a good idea. We have already this file named `get_block_template.rs`. I am thinking on renaming it to `get_block_template_rpcs.rs` and then do `get_block_template.rs` to hold `GetBlockTemplate`. 

Then `default_roots.rs` for `DefaultRoots`, `transaction.rs` for `Transaction` and `coinbase.rs` for `Coinbase`.

Let me know if something like that is what you have in mind or something different.

Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template.rs]: Another option can be to create a `types` module/folder and drop the type files inside it.

Comment (arya2) [reply zebra-rpc/src/methods/get_block_template.rs]: I had a file for each method in mind, i.e. `methods/types/getblocktemplate.rs`, but on second thought that would be confusing, and I like your suggestion better.

**Review by teor2345 — APPROVED** [review]
Looks good!

I think this is a high priority, because it renames the module that all the other RPCs are in. If we don't get it merged soon, it will cause code conflicts.

**Pull request merged**